public/app/features/dashboard/state/PanelModel.ts

```diff
@@ -269,6 +270,33 @@ export class PanelModel implements DataConfigSource, IPanelModel {
     }
   }
 
+  // Attempts to populate each panel target with a well-formed query definition
+  async ensureWellFormedQueries() {
+    if (this.targets && isArray(this.targets)) {
+      // If we modified any targets, publish one update event
+      let shouldPublish = false;
+      for (let i = 0; i < this.targets.length; i++) {
+        const query = this.targets[i];
+        if (!query.datasource && this.datasource) {
+          query.datasource = this.datasource;
+        }
+        if (Object.keys(query).length === 2) {
```

Comment:
This is based on the combination of the above if statement, plus the `ensureQueryIds()` function. If the only two properties on the query are the ones we are hard-coding in, the query is not well-formed. If there is a better way of inferring this, let me know.